### PR TITLE
Update CSR docs with expirationSeconds field details

### PR DESCRIPTION
**Pull request by enj**
Signed-off-by: Monis Khan <mok@vmware.com>

xref https://github.com/kubernetes/kubernetes/pull/99494
xref https://github.com/kubernetes/enhancements/pull/2759

/assign @mikedanese @liggitt

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
@liggitt @mikedanese @munnerz can you confirm you're happy to add @enj as reviewer to this page?

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:96]:
```suggestion
The `spec.durationHint` field was added in Kubernetes v1.22. Earlier versions of Kubernetes do not honor this field.
```
What happens if you try to set that field in a CSR submitted to an older API server?

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:159]:
```suggestion
The `spec.durationHint` field was added in Kubernetes v1.22.  Earlier versions of Kubernetes do not honor this field.
```

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:399]:
```suggestion
The `spec.durationHint` field was added in Kubernetes v1.22.  Earlier versions of Kubernetes do not honor this field.
```

**Review by sftim — CHANGES_REQUESTED** [review]
This change is invalid against `master` I'm afraid (`master` is the published documentation, currently v1.22).

I think you meant to propose a change to the _dev-1.22_ branch.

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:36]: For this first mention of the field:
- I'd mention it's only available in Kubernetes v1.22 and later
- I'd also tell people what the type of the field is. They might for example think it takes a number of seconds or a number of days, or more likely be frustrated because both could be valid and they're not sure.

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (enj) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:64]: The existing docs are incorrect and incomplete.

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (enj) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:96]: Like all new fields added in newer Kubernetes versions, older API servers will silently drop the field.

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:36]: yes to noting the version this is supported in

API convention is to include the unit in the field name and make it an integer, not a string (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units)

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:64]: I'd suggest splitting the corrections to current docs to a PR to `master` and merging it now

**Comment by liggitt** [comment]
> @liggitt @mikedanese @munnerz can you confirm you're happy to add @enj as reviewer to this page?

+1

**Comment by liggitt** [comment]
/hold for https://github.com/kubernetes/kubernetes/pull/99494

I like the direction, should hold until the PR merges and the field name / semantics are finalized

**Comment by PI-Victor** [comment]
Hi @liggitt i'm a bit confused about this change, it doesn't look like it has opted to be in the 1.22 release, i can't find it anywhere in the tracking sheet? I also see it's non-trivial change

**Comment by kikisdeliveryservice** [comment]
For ref: https://github.com/kubernetes/enhancements/pull/2759#issuecomment-848156775

This shouldn't merge until this change has an active kep and will need an [exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) for inclusion in 1.22.

**Comment by sftim** [comment]
/hold
(these aren't reentrant locks - but we should also wait for the KEP)

**Comment by kikisdeliveryservice** [comment]
As a note the author received an exception request and the required [KEP has merged](https://github.com/kubernetes/enhancements/pull/2788)

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:36]: I guess I'd like this to be called `durationSecondsHint` or similar, then (and be an integer).

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:96]: > Like all new fields added in newer Kubernetes versions, older API servers will silently drop the field.

I recommend explaining that in the text: we shouldn't assume readers are closely familiar with Kubernetes API versioning rules.

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:83]: nit: PKCS#10 as a format does allow specifying this (via an extension); the problem is that there's not standard OID for recording the expiration date or lifetime. (I think we're saying that this separate mechanism makes more sense for Kubernetes than requiring users to learn how to use some custom CNCF OID for that, with all the ASN.1 fun that would entail - and I agree!

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:36]: If the field name is already decided, then instead we must document what units it uses. Ideally the API reference for CertificateSigningRequest will use the same, clear working.

Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:36]: final field name is expirationSeconds

File: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:35]: It seems like this needs to change to reflect the field renaming to `.spec.expirationSeconds`

**Comment by PI-Victor** [comment]
@enj can you please rebase and address the requested changes? thank you!

**Comment by enj** [comment]
@PI-Victor @sftim comments addressed.

The first commit is #29042.

**Comment by sftim** [comment]
@kubernetes/sig-auth-pr-reviews does this look OK from a technical perspective?

**Comment by liggitt** [comment]
tech content lgtm

**Comment by tengqm** [comment]
Glad to see this one is ready.
/lgtm
/approve

**Pull request merged**
